core/src/main/java/org/elasticsearch/index/query/functionscore/DecayFunctionBuilder.java

```diff
@@ -88,4 +127,372 @@ public ScoreFunctionBuilder setMultiValueMode(String multiValueMode) {
         this.multiValueMode = MultiValueMode.fromString(multiValueMode.toUpperCase(Locale.ROOT));
         return this;
     }
+
+    @Override
+    public QueryValidationException validate() {
+        QueryValidationException validationException = null;
+        if (fieldName == null) {
+            validationException = addValidationError("field name cannot be null", validationException);
+        }
+        if (scale == null) {
+            validationException = addValidationError("scale must not be null", validationException);
+        }
+        if (decay <= 0 || decay >= 1.0) {
+            validationException = addValidationError("decay must be in range 0..1!", validationException);
+        }
+        return validationException;
+    }
+
+    @Override
+    protected ScoreFunction doScoreFunction(QueryShardContext parseContext) throws IOException {
+        // now, the field must exist, else we cannot read the value for the doc later
+        MappedFieldType fieldType = parseContext.fieldMapper(fieldName);
+        if (fieldType == null) {
+            throw new IllegalArgumentException("unknown field ["+fieldName+"]");
+        }
+        // dates and time need special handling
+        if (fieldType instanceof DateFieldMapper.DateFieldType) {
+            return handleDateVariable(parseContext, (DateFieldMapper.DateFieldType) fieldType);
+        } else if (fieldType instanceof GeoPointFieldMapper.GeoPointFieldType) {
+            return handleGeoVariable(parseContext, (GeoPointFieldMapper.GeoPointFieldType) fieldType);
+        } else if (fieldType instanceof NumberFieldMapper.NumberFieldType) {
+            return handleNumberVariable(parseContext, (NumberFieldMapper.NumberFieldType) fieldType);
+        } else {
+            throw new IllegalArgumentException("field ["+fieldName+"] is of type ["+fieldType+"], but only numeric types are supported.");
+        }
+    }
+
+    private AbstractDistanceScoreFunction handleNumberVariable(
+            QueryShardContext parseContext, NumberFieldMapper.NumberFieldType fieldType) throws IOException {
+        // to be consistent with the parser
+        if (this.scale == null || this.origin == null) {
+            throw new IllegalArgumentException("both ["+DecayFunctionBuilder.SCALE+"] and ["+DecayFunctionBuilder.ORIGIN+"] " +
+                    "must be set for numeric fields.");
```

Comment:
If I'm not mistaken in other similar instances we throw a QueryShardException if an expected field is not found. If so, should we do that here as well?

See https://github.com/elastic/elasticsearch/pull/13393/files#diff-78e5ba6228b6e21d73de747f33eca3baR198 for comparison.